core/controllers/base_test.py

```diff
@@ -274,6 +288,179 @@ def test_renders_error_page_with_iframed(self):
+        # would just call the existing module.
+        del sys.modules['feconf']
+        server_software_swap = self.swap(
+            os, 'environ', {'SERVER_SOFTWARE': 'Production'})
+        assert_raises_regexp_context_manager = self.assertRaisesRegexp(
+            Exception, 'DEV_MODE can\'t be true on production.')
+        with assert_raises_regexp_context_manager, server_software_swap:
+            # This pragma is needed since we are re-importing under
+            # invalid conditions. The pylint error messages
+            # 'reimported', 'unused-variable', 'redefined-outer-name' and
+            # 'unused-import' would appear if this line was not disabled.
+            import feconf  # pylint: disable-all
+
+    def test_valid_pillow_path(self):
+        # We need to re-import appengine_config here to make it look like a
+        # local variable so that we can again re-import appengine_config later.
+        import appengine_config
+        assert_raises_regexp_context_manager = self.assertRaisesRegexp(
+            Exception, 'Invalid path for oppia_tools library: invalid_path')
+
+        def mock_os_path_join_for_pillow(*args):
+            """Mocks path for 'Pillow' with an invalid path. This is done by
+            substituting os.path.join to return an invalid path. This is needed
+            to test against an invalid path for 'Pillow'.
+            """
+            path = ''
+            if args[1] == 'Pillow-6.0.0':
+                return 'invalid_path'
+            else:
+                path = '/'.join(args)
+                return path
+
+        pil_path_swap = self.swap(os.path, 'join', mock_os_path_join_for_pillow)
+        # We need to delete the existing module else the re-importing
+        # would just call the existing module.
+        del sys.modules['appengine_config']
+
+        with assert_raises_regexp_context_manager, pil_path_swap:
+            # This pragma is needed since we are re-importing under
+            # invalid conditions. The pylint error messages
+            # 'reimported', 'unused-variable', 'redefined-outer-name' and
+            # 'unused-import' would appear if this line was not disabled.
+            import appengine_config  # pylint: disable-all
+
+    def test_valid_third_party_library_path(self):
+        # We need to re-import appengine_config here to make it look like a
+        # local variable so that we can again re-import appengine_config later.
+        import appengine_config
+        assert_raises_regexp_context_manager = self.assertRaisesRegexp(
+            Exception, 'Invalid path for third_party library: invalid_path')
+
+        def mock_os_path_join_for_third_party_lib(*args):
+            """Mocks path for third_party libs with an invalid path. This is
+            done by substituting os.path.join to return an invalid path. This is
+            needed to test against an invalid path for third_party libs.
```

Comment:
This is needed to test the scenario where the third_party libs path points to a non-existent directory.

That said, actually, it is a bit weird to swap out such core functionality like os.path.join here, especially with an implementation that's not generic. Why not swap out appengine_config.THIRD_PARTY_LIBS instead (with a list with one item pointing to a non-existent directory)?

Reply:
Same note for mock_os_path_join_for_pillow.

Reply:
> Why not swap out appengine_config.THIRD_PARTY_LIBS instead (with a list with one item pointing to a non-existent directory)?

If we swap appengine_config.THIRD_PARTY_LIBS and again reimport appengine_config the existing swap will be lost since before reimporting we would need to do `del sys.modules`

Reply:
Fixed the comments